### PR TITLE
TG-127: Superadmin Form Validations and enrouting with React Lazy

### PR DESCRIPTION
**Pull request by fernandocarbone1**
1. JOI validations
2. React Hook Form
3. React Lazy for routes

### CODE REVIEW

File: src/Components/SuperAdmins/index.jsx [hunk 1]
```diff
@@ -119,59 +149,68 @@ function SuperAdmins() {
...
-              onChange={(e) => setFirstName(e.target.value)}
-            />
+            <label htmlFor="firstName">First Name</label>
+            <input {...register('firstName')} name="firstName" type="text" placeholder="John" />
+            {errors.firstName?.type === 'string.empty' && (
+              <p className={styles.error}>{errors.firstName.message}</p>
+            )}
+            {errors.firstName?.type === 'string.min' && (
+              <p className={styles.error}>{errors.firstName.message}</p>
+            )}
```
Comment (martinalbrnz) [inline src/Components/SuperAdmins/index.jsx:166]: faltan validaciones de .max() y podria haber tambien de regex (que solo acepte letras y espacios)

File: src/Components/SuperAdmins/index.jsx [hunk 2]
```diff
@@ -119,59 +149,68 @@ function SuperAdmins() {
...
-              onChange={(e) => setLastName(e.target.value)}
-            />
+            <label htmlFor="lastName">Last Name</label>
+            <input {...register('lastName')} name="lastName" type="text" placeholder="Doe" />
+            {errors.lastName?.type === 'string.empty' && (
+              <p className={styles.error}>{errors.lastName.message}</p>
+            )}
+            {errors.lastName?.type === 'string.min' && (
+              <p className={styles.error}>{errors.lastName.message}</p>
+            )}
```
Comment (martinalbrnz) [inline src/Components/SuperAdmins/index.jsx:176]: Idem validaciones, revisalas en general. Otra alternativa es que controles los mensajes desde el schema de Joi

File: src/Components/SuperAdmins/index.jsx [hunk 3]
```diff
@@ -30,31 +50,26 @@ function SuperAdmins() {
   const modalShowForm = useSelector((state) => state.superAdmins.showFormAddEdit);
   const showModalDelete = useSelector((state) => state.superAdmins.showModalDelete);
   const showModalMessage = useSelector((state) => state.superAdmins.showModalMessage);
-  const [firstName, setFirstName] = useState('');
-  const [lastName, setLastName] = useState('');
-  const [email, setEmail] = useState('');
-  const [password, setPassword] = useState('');
-  const [active, setActive] = useState(true);
+  // const [active, setActive] = useState(true);
```
Comment (nicolascostanza) [inline src/Components/SuperAdmins/index.jsx:53]: este comment borralo

File: src/Components/SuperAdmins/index.jsx [hunk 4]
```diff
@@ -30,31 +50,26 @@ function SuperAdmins() {
...
-    setPassword(valuesForm[0].password);
-    setActive(valuesForm[0].active === 'true' ? true : false);
+    const valuesForm = superAdminsList.find((superadmin) => superadmin._id === id);
+    reset({
+      firstName: valuesForm.firstName,
+      lastName: valuesForm.lastName,
+      email: valuesForm.email,
+      password: valuesForm.password,
+      active: valuesForm.active === 'true' ? true : false
+    });
```
Comment (nicolascostanza) [inline src/Components/SuperAdmins/index.jsx:72]: nice

File: src/Components/SuperAdmins/index.jsx [hunk 5]
```diff
@@ -80,15 +95,16 @@ function SuperAdmins() {
...
-      email,
-      password,
-      active
+      firstName: data.firstName,
+      lastName: data.lastName,
+      email: data.email,
+      password: data.password,
+      active: Boolean(data.active)
     };
+    console.log(data);
```
Comment (nicolascostanza) [inline src/Components/SuperAdmins/index.jsx:107]: este console.log fuera

File: src/Components/SuperAdmins/index.jsx [hunk 6]
```diff
@@ -97,6 +113,20 @@ function SuperAdmins() {
       alert('Something unexpected happened');
     }
   };
+  const {
+    register,
+    handleSubmit,
+    reset,
+    getValues,
+    // setValue,
```
Comment (nicolascostanza) [inline src/Components/SuperAdmins/index.jsx:121]: este comment no deberia estar

File: src/Components/SuperAdmins/index.jsx [hunk 7]
```diff
@@ -97,6 +113,20 @@ function SuperAdmins() {
...
+    formState: { errors }
+  } = useForm({
+    mode: 'onBlur',
+    resolver: joiResolver(schema)
+  });
+  const formData = getValues('active');
+  console.log(formData);
+  // console.log(errors);
```
[the remark this answers is not in the document]
Comment (nicolascostanza) [reply src/Components/SuperAdmins/index.jsx:129]: x2

File: src/Components/SuperAdmins/index.jsx (hunk 2 above)
Comment (nicolascostanza) [reply src/Components/SuperAdmins/index.jsx:176]: x2, mejor desde el schema

File: src/Components/SuperAdmins/index.jsx [hunk 8]
```diff
@@ -119,59 +149,68 @@ function SuperAdmins() {
...
               <input
-                className={styles.inputsDivs}
                 type="checkbox"
-                checked={active}
-                value={active}
-                onChange={(e) => setActive(e.currentTarget.checked)}
+                className={styles.inputsDivs}
+                {...register('active')}
+                // checked={!formData}
+                // onChange={setValue('active', getValues('active'))}
```
Comment (nicolascostanza) [inline src/Components/SuperAdmins/index.jsx:213]: estos comments nop

File: src/Components/SuperAdmins/index.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (FefeTroanes) [reply src/Components/SuperAdmins/index.jsx:129]: x3 (?

File: src/Components/SuperAdmins/index.jsx [hunk 9]
```diff
@@ -119,60 +172,49 @@ function SuperAdmins() {
...
           <div className={styles.inputsForm}>
-            <label>Password</label>
-            <input
-              className={styles.inputsDivs}
-              type="password"
-              value={password}
-              onChange={(e) => setPassword(e.target.value)}
-            />
+            <label htmlFor="password">Password</label>
+            <input {...register('password')} type="password" />
```
Comment (martinalbrnz) [inline src/Components/SuperAdmins/index.jsx:207]: missing password placeholder

File: src/Components/SuperAdmins/index.jsx [hunk 10]
```diff
@@ -22,6 +25,53 @@ function SuperAdmins() {
...
   const dispatch = useDispatch();
+  const schema = Joi.object({
+    firstName: Joi.string()
+      .min(3)
+      .max(50)
+      .required()
+      .regex(/^([ \u00c0-\u01ffa-zA-Z'-])+$/)
+      .messages({
+        'string.min': 'First name must contain at least 3 characters',
+        'string.max': 'First name must contain less than 30 characters',
```
Comment (martinalbrnz) [inline src/Components/SuperAdmins/index.jsx:36]: .max() have 50
string.max say 30

File: src/Components/SuperAdmins/index.jsx [hunk 11]
```diff
@@ -22,6 +25,53 @@ function SuperAdmins() {
...
+      .max(50)
+      .required()
+      .lowercase()
+      .regex(
+        // eslint-disable-next-line no-useless-escape
+        /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
+      )
+      .messages({
+        'string.min': 'Email must contain at least 3 characters',
+        'string.max': 'Email must contain less than 30 characters',
```
Comment (martinalbrnz) [inline src/Components/SuperAdmins/index.jsx:62]: .max() have 50
string.max say 30

File: src/Components/SuperAdmins/index.jsx [hunk 12]
```diff
@@ -22,6 +25,53 @@ function SuperAdmins() {
...
+        'string.pattern.base': 'Email must be valid',
+        'string.empty': 'This field is required'
+      }),
+    password: Joi.string().min(3).max(50).required().messages({
+      'string.min': 'Password must contain at least 3 characters',
+      'string.max': 'Password must contain less than 30 characters',
+      'string.pattern.base': 'Password must be valid',
+      'string.empty': 'This field is required'
```
Comment (martinalbrnz) [inline src/Components/SuperAdmins/index.jsx:70]: you have no pattern to validate password, string.pattern.base will never trigger
also, min() should be 8.
Password should have numbers and letters

**Review by JuanBSantomero — DISMISSED** [review]
Everything seems good to go.

**Review by mpueblas — COMMENTED** [review]
yo lo veo bien, pero fijate que vercel no te lo está aprobando

**Review by mpueblas — DISMISSED** [review]

**Comment by karu-ams** [comment]
## TEST RESULTS
Tested locally on https://alfonso-trackgenix-otcu15jym-basp-m2022.vercel.app/

Test Results: FAILED ⚠️ 

BUGS FOUND:
1. When creating a superadmin, I used "1234asdf" as a password and I received an error when it should pass the validation 🤔 
image

2. SUGGESTION: Maybe the state of the checkbox of the superadmin could be ACTIVE by default. ☑️ 
image

3. I tried to edit an old user, when I click on the password the validation failed, depite the fact I didn't change a thing.
image

**Review by GinaSchP — APPROVED** [review]

**Review by GinaSchP — APPROVED** [review]

**Comment by karu-ams** [comment]
## TEST RESULTS
Tested locally on https://alfonso-trackgenix-3qoszn0v0-basp-m2022.vercel.app/

Test Results: PASS ✅ READY TO MERGE

**Review by karu-ams — APPROVED** [review]
## TEST RESULTS
_Tested locally on https://alfonso-trackgenix-3qoszn0v0-basp-m2022.vercel.app/_

**Test Results: PASS ✅ READY TO MERGE**

**Pull request merged**